src/DotNetty.Common/ThreadLocalPool.cs

```diff
@@ -41,21 +50,307 @@ public bool Release<T>(T value)
             }
         }
 
-        internal sealed class Stack : Stack<Handle>
+        internal sealed class WeakOrderQueue
         {
-            public readonly ThreadLocalPool Owner;
+            const int LINK_CAPACITY = 16;
+            class Link
+            {                
+                internal int readIndex;
+                internal volatile int writeIndex;
+
+                internal Handle[] elements;
+                internal Link next;
+
+                internal Link()
+                {
+                    elements = new Handle[LINK_CAPACITY];
+                }
+
+                internal void Add(Handle handle)
+                {
+                    elements[writeIndex++] = handle;
+                }
+
+                internal bool CanAdd
+                {
+                    get
+                    {
+                        return (writeIndex < LINK_CAPACITY - 1);
+                    }
+                }
+
+                internal bool CanRead
+                {
+                    get
+                    {
+                        return readIndex < elements.Length;
+                    }
+                }
+            }
+
+            Link head, tail;
+            internal WeakOrderQueue next;
+            internal WeakReference<Thread> ownerThread;
+
+            internal bool IsEmpty()
+            {
+                return tail.readIndex == tail.writeIndex;
+            }
+
+            internal WeakOrderQueue(Stack stack, Thread thread)
+            {
+                Contract.Requires(stack != null);
+
+                ownerThread = new WeakReference<Thread>(thread);
+                head = tail = new Link();
+                lock(stack) {
+                    next = stack.headQueue;
+                    stack.headQueue = this;
+                }
+            }
+
+            internal void Add(Handle handle)
+            {
+                Contract.Requires(handle != null);
+
+                //handle.lastRecycledId = id;
+
+                Link tail = this.tail;
+                int writeIndex;
+                if ((writeIndex = tail.writeIndex) == LINK_CAPACITY)
+                {
+                    this.tail = tail = tail.next = new Link();
+                    writeIndex = tail.writeIndex;
+                }
+                tail.elements[writeIndex] = handle;
+                //handle.Stack = null;
+                tail.writeIndex++;
```

Comment:
in another comment I'm asking to replace `volatile` keyword with explicit use of `Volatile.*` methods. I suppose it got lost because of a new commit. Check this for details: http://blog.alexrp.com/2014/03/30/dot-net-atomics-and-memory-model-semantics/